xarray/backends/api.py

```diff
@@ -80,6 +80,28 @@
 }
 
 
+def avail_engines() -> List:
```

Comment:
You can type hint the contents of the `List` here too, for example if you choose to go with my other suggestion it would be something like

```python
T_Backend = TypeVar('T_Backend', bound=BackendEntrypoint)

def avail_engines() -> List[T_Backend]:
```

Here `T_Backend` is type hinted in a way that means "an instance of the `BackendEntrypoint` class or an instance of any class which subclasses `BackendEntrypoint`."

(Apologies if you already know this stuff!)

Reply:
Thanks - this is super helpful! If I'm grasping correctly though, each element of the list isn't actually a `T_Backend` type, because the list is custom built to have certain info FROM each of the class instances. Similarly, with the current version of the function, `def avail_engines() -> pd.DataFrame` is as specific as we can get because a dataframe holds multiple types of data...

Reply:
Yes I think you're grasping correctly. 

> Similarly, with the current version of the function, def avail_engines() -> pd.DataFrame is as specific as we can get because a dataframe holds multiple types of data...

I think that is true unfortunately. (As I don't know of a way to type hint the contents of the `pd.DataFrame`.)

Reply:
@JessicaS11 we know enough about the design to type hint this now, it would be

```python
T_Backend = TypeVar('T_Backend', bound=BackendEntrypoint)  # this should possibly live in `xarray.core.types.py`

def list_engines() -> Mapping[str, T_Backend]:
    ...
```